HolidayViewset 작성 && Schedule 나가기, 삭제 로직 삭정

cafitac opened the pull request:
schedule/serializers.py
 - ADD: HolidaySerializer 생성

schedule/views.py
 - ADD: 리스트만 불러올 수 있는 HolidayViewSet 생성
 - ADD: year로 필터링할 수 있는 filter action 추가
 - ADD: ScheduleViewSet에 일정 등록자라면 삭제, 참가자라면 나가기 기능을 수행하는 leave action 추가

schedule/urls.py
 - router에 HolidayViewSet 등록

Commit:
HolidayViewset 작성

schedule/serializers.py
 - ADD: HolidaySerializer 생성

schedule/views.py
 - ADD: 리스트만 불러올 수 있는 HolidayViewSet 생성
 - ADD: year로 필터링할 수 있는 filter action 추가

schedule/urls.py
 - router에 HolidayViewSet 등록

Commit:
Schedule 나가기, 삭제 로직 작성

schedule/views.py
 - ADD: ScheduleViewSet에 일정 등록자라면 삭제, 참가자라면 나가기 기능을 수행하는 leave action 추가

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -59,4 +59,17 @@ def join(self, request, pk=None):
+
+
+class HolidayViewSet(viewsets.GenericViewSet, mixins.ListModelMixin):
+    queryset = Holiday.objects.all().filter(is_holiday=True)
```
all() 필요없음

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -59,4 +59,17 @@ def join(self, request, pk=None):
+    @action(detail=False, methods=['GET'])
+    def filter(self, request):
+        year = request.GET.get('year')
+        holidays = Holiday.objects.all().filter(is_holiday=True, date__year=year)
```
1. 여기도 마찬가지
2. year 파라미터 없을경우 필터 어떻게 먹힐까

cafitac replied:
아하... year 파라미터 없을 때는 테스트를 깜빡하고 안했네요 ㅠㅠ

Commit:
자신 Schedule 확인하는 action 추가

schedule/serializers.py
 - MOD: 기존 ( ScheduleSerializer -> ScheduleCreateSerializer )로 이름 변경
 - ADD: ScheduleListSerializer 추가

schedule/views.py
 - ADD: ScheduleViewSet에 자신의 일정을 확인할 수 있는 filter action 추가

Commit:
공휴일 제공 방식 변경

schedule/views.py
 - DEL: 년도 별로 제공하는 filter action 뷰 제거
 - ADD: get 파라미터가 없다면 현재 db에 있는 모든 년도의 공휴일을 제고하고 있다면 해당 년도만 제공하도록 변경

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -61,6 +60,54 @@ def join(self, request, pk=None):
+                }, status=status.HTTP_400_BAD_REQUEST)
+            participant = User.objects.get(pk=participant_id)
+            if participant.id == schedule.registrant_id:
+                Schedule.objects.filter(pk=pk).delete()
```
schedule의 pk값은 고유값이 아닌가 ? 고유값이면 get 을 써야할거같은데

cafitac replied:
get으로 하면 delete가 한 번에 되지가 않아서 인스턴스를 받은 다음에 delete 작업을 해야된다고 하던데 그냥 두 줄로 늘려서 get하는 방식으로 바꿀까요??

hanseongd replied:
에? 될텐데 ?

cafitac replied:
헐 그렇네요.. 스택오버플로우에서 안된다고 해서 filter로 했던건데...
get으로 수정할게욥

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -61,6 +60,54 @@ def join(self, request, pk=None):
 
+    @action(detail=True, methods=['GET'])
+    def leave(self, request, pk=None):
+        schedule = Schedule.objects.get(pk=pk)
```
pk가 없다면?

cafitac replied:
action 뷰에서는 pk가 필수로 있어야 api를 호출할 수 있어서 그냥 이상태로 코드 유지해도 될 것 같은데 아닌가요??

hanseongd replied:
url 형식만 맞으면 들어오지 않아 ? 호출이 안될 경우 없다는 message가 날아갈텐데 해당 형식이 우리가 정한 형식하고 맞는지도 모르겠고

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -61,6 +60,54 @@ def join(self, request, pk=None):
+        participant_id = request.GET.get('id')
+        try:
+            try:
+                Schedule.objects.get(pk=pk, participants=participant_id)
```
schedule = Schedule.objects.get(pk=pk) 이 부분과 Schedule.objects.get(pk=pk, participants=participant_id) 무슨 차이야?

cafitac replied:
위에서 pk로만 필터링하는 건 말 그대로 schedule만 가져오는거고 아래에서 participant_id로 필터링하는 건 그 일정에 유저가 참가하고 있는지 여부를 알려고 한건데 두 개를 같이하면 일정이 없는 건지, 유저가 일정에 참가하고 있지 않은건지 알 수가 없더라고요... 그래서 두 개로 나눠서 처리했어욤..

hanseongd replied:
schedule = Schedule.objects.get(pk=pk) 이렇게 가져온거에서
schedule. participants.id == participant_id 이런식으로 판별하는건?

cafitac replied:
아하 그 방법이 있네요.. 생각을 못했습니다

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -108,6 +108,28 @@ def leave(self, request, pk=None):
 
+    @action(detail=False, methods=['GET'])
+    def filter(self, request):
+        user_id = request.GET.get('id')
```
id가 없을 경우 바로 return 해버리면 query 안날리고 리소스를 줄일 수 있을 것 같은데

cafitac replied:
저 id는 필수로 받아야되는 것 같아요. id로 유저 아이디로 가져와서 그 유저가 참가하고 있는 일정이 있는지 확인하는 것라서....

hanseongd replied:
파라미터에 ?id= 이런식으로 요청을 해버리면 빈값으로 받을텐데 이상태로 filter를 해버리면 참여 중인 일정이 없다고 리턴이 될거고 그럼 잘못된 요청임을 인지 못하는 상황이 발생할 수 있지 않을까

cafitac replied:
넵넵 exception 추가할게요

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -132,13 +132,21 @@ def filter(self, request):
-        holidays = Holiday.objects.all().filter(is_holiday=True, date__year=year)
-        serializer = HolidaySerializer(holidays, many=True)
 
+        if year:
```
```
queryset = self.filter_queryset(self.get_queryset())
if year:
    queryset = queryset.filter(date__year=year)
```

cafitac replied:
아 이렇게 했는데 queryset이 두 번 저장되는 줄 알고 다시 바꿨었는데...ㅠㅠㅠ
바꿀게욤

Commit:
일정 유저 join, leave 로직 변경

schedule/views.py
 - MOD: [ScheduleViewSet] join action 뷰 response 구조 변경
 - ADD: [ScheduleViewSet] join action 뷰 querystring에 대한 exception 추가
 - MOD: [ScheduleViewSet] leave action 뷰 일정에 유저 참가 중인지 확인하는 로직 변경
 - ADD: [ScheduleViewSet] leave action 뷰 querystring에 대한 exception 추가
 - MOD: [HolidayViewSet] list 함수, 년도로 필터링하는 로직 변경

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -38,45 +38,59 @@ def create(self, request, *args, **kwargs):
+            participant_id = request.GET.get('id')
+            schedule = Schedule.objects.get(pk=pk)
+
+            participants_ids = [participant.pk for participant in schedule.participants.all()]
```
```
if not schedule.participants.filter(pk=participant_id).exist():
    return Response({
        'success': False,
        'data': {
            'message': '해당 일정에 참가하고 있지 않습니다.'
        }
    }, status=status.HTTP_400_BAD_REQUEST)
```
이런식으로 하면 될거같은데

cafitac replied:
넵 확인했어요 exist()는 못 쓸 것 같고 filter로만 판단하는 걸로 고칠 수 있을 것 같아여

hanseongd replied:
exist() 못써 ?

leensungzero replied:
네 filter로 가져오는거라 안되는 것 같아요. get으로도 해봤는데 MTM field에서 get은 지원을 안 해주는 것 같구요

cafitac replied:
아 제가 착각했었네요...

Commit:
일정 멤버 추가 로직 변경

schedule/views.py
 - MOD: 일정에 멤버가 참가하고 있는지에 대한 여부를 판단하는 로직 변경

Commit:
일정 멤버 추가 로직 개선

schedule/views.py
 - MOD: 일정에 멤버가 참가하고 있는지에 대한 여부를 exists()를 이용하여 판단

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -38,25 +38,134 @@ def create(self, request, *args, **kwargs):
+                'data': {
+                    'message': '해당 유저를 찾을 수 없습니다.'
+                }
+            }, status=status.HTTP_400_BAD_REQUEST)
```
400 이랑 밑 value error의 404 status code 반대로 쓴거 같은데??

cafitac replied:
아하 어떤걸 400주고 어떤걸 404 줄지 헷갈려서 일단 저렇게 줬었는데 지금 보니 바꾸는게 맞네욥

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -38,25 +38,134 @@ def create(self, request, *args, **kwargs):
+                'data': {
+                    'message': '해당 일정을 찾을 수 없습니다.'
+                }
+            }, status=status.HTTP_400_BAD_REQUEST)
```
여기도

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -38,25 +38,134 @@ def create(self, request, *args, **kwargs):
+                "data": {
+                    "message": "참여 중인 일정이 없습니다."
+                }
+            })
```
status code 200이라서 일부러 안 넣은건지

cafitac replied:
완전한 제 실수입니다...ㅠㅠ

hanseongd reviewed (commented):
이 PR의 마지막 리뷰인듯

Commit:
status code 용도에 맞게 다시 변경

schedule/views.py
 - MOD: 상황에 맞는 status code를 제공하도록 변경

hanseongd commented on scheduler/schedule/views.py:
```diff
@@ -38,25 +38,134 @@ def create(self, request, *args, **kwargs):
+
+            participant = User.objects.get(pk=participant_id)
+            if participant.id == schedule.registrant_id:
+                Schedule.objects.get(pk=pk).delete()
```
schedule = Schedule.objects.get(pk=pk) 이 부분이 있으니까
Schedule.objects.get(pk=pk).delete() -> schedule.delete()
한 번에 못봐줘서 미안 ㅠㅠ

cafitac replied:
아 위에 이미 선언해놨었네요.. 죄송합니다...

Commit:
사소한 로직 수정

schedule/views.py
 - MOD: Schedule 인스턴스를 새로 가져오지 않고 위에서 선언한 schedule 인스턴스를 재활용하도록 변경
